lib/src/test/java/io/ably/lib/http/HttpTest.java

```diff
@@ -0,0 +1,447 @@
+package io.ably.lib.http;
+
+import fi.iki.elonen.NanoHTTPD;
+import fi.iki.elonen.router.RouterNanoHTTPD;
+import io.ably.lib.test.util.StatusHandler;
+import io.ably.lib.types.AblyException;
+import io.ably.lib.types.ClientOptions;
+import io.ably.lib.types.ErrorInfo;
+import io.ably.lib.types.Param;
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeMatcher;
+import org.junit.*;
+import org.junit.rules.ExpectedException;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.*;
+
+import static org.hamcrest.Matchers.*;
+import static org.junit.Assert.assertThat;
+import static org.mockito.AdditionalMatchers.aryEq;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyBoolean;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.*;
+
+/**
+ * Created by gokhanbarisaker on 2/2/16.
+ */
+public class HttpTest {
+
+    private static final String PATTERN_HOST_FALLBACK = "(?i)[a-e]\\.ably-realtime.com";
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+    private static RouterNanoHTTPD server;
+
+
+    @BeforeClass
+    public static void setUp() throws IOException {
+        server = new RouterNanoHTTPD(27331);
+        server.addRoute("/status/:code", StatusHandler.class);
+        server.start(NanoHTTPD.SOCKET_READ_TIMEOUT, true);
+
+        while (!server.wasStarted()) {
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    @AfterClass
+    public static void tearDown() {
+        server.stop();
+    }
+
+
+    /*******************************************
+     * Spec: RSC15
+     *******************************************/
+
+
+    /**
+     * <p>
+     * Validates http is not using any fallback host when we receive valid response from http's host
+     * </p>
+     * <p>
+     * Spec: RSC15a
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_nofallback() throws Exception {
+        Http http = Mockito.spy(new Http(new ClientOptions(), null));
+
+        String hostExpected = "non.fallback.host";
+        String responseExpected = "Meow";
+        ArgumentCaptor<URL> argument = ArgumentCaptor.forClass(URL.class);
+        http.setHost(hostExpected);
+
+        /* Partially mock http */
+        doReturn(responseExpected)
```

Comment:
I do like unit tests, so I would prefer just one or two that hit the network to prove that the library is doing what it should do, and these unit tests to quickly test the permutations as you are